instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshStateTest.kt

```diff
@@ -0,0 +1,547 @@
+import com.mapbox.navigation.instrumentation_tests.utils.http.NthAttemptHandler
+import com.mapbox.navigation.instrumentation_tests.utils.readRawFileText
+import com.mapbox.navigation.testing.ui.BaseTest
+import com.mapbox.navigation.testing.ui.http.MockRequestHandler
+import com.mapbox.navigation.testing.ui.utils.getMapboxAccessTokenFromResources
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.take
+import kotlinx.coroutines.flow.toList
+import org.junit.Assert.assertEquals
+import org.junit.Rule
+import org.junit.Test
+import java.net.URI
+import java.util.concurrent.TimeUnit
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+class RouteRefreshStateTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.java) {
+
+    @get:Rule
+    val mapboxNavigationRule = MapboxNavigationRule()
+
+    private lateinit var mapboxNavigation: MapboxNavigation
+    private val observer = TestObserver()
+
+    private val coordinates = listOf(
+        Point.fromLngLat(-121.496066, 38.577764),
+        Point.fromLngLat(-121.480279, 38.57674)
+    )
+
+    override fun setupMockLocation(): Location = mockLocationUpdatesRule.generateLocationUpdate {
+        latitude = coordinates[0].latitude()
+        longitude = coordinates[0].longitude()
+    }
+
+    @Test
+    fun emptyRoutesOnDemandRefreshTest() = sdkTest {
+        setupMockRequestHandlers(
+            coordinates,
+            R.raw.route_response_single_route_refresh,
+            createRefreshHandler(
+                R.raw.route_response_route_refresh_annotations,
+                "route_response_single_route_refresh"
+            )
+        )
+
+        createMapboxNavigation(createRouteRefreshOptionsWithInvalidInterval(1_000))
+        mapboxNavigation.startTripSession()
+        val requestedRoutes = requestRoutes()
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(requestedRoutes)
+        mapboxNavigation.clearNavigationRoutesAndWaitForUpdate()
+        delay(2000) // refresh interval + accuracy
+        mapboxNavigation.registerRouteRefreshStateObserver(observer)
+        mapboxNavigation.refreshRoutesImmediately()
+
+        assertEquals(emptyList<String>(), observer.getStatesSnapshot())
+    }
+
+    @Test
+    fun routeRefreshOnDemandForInvalidRoutes() = sdkTest {
+        setupMockRequestHandlers(
+            coordinates,
+            R.raw.route_response_single_route_refresh,
+            createRefreshHandler(
+                R.raw.route_response_route_refresh_annotations,
+                "route_response_single_route_refresh"
+            )
+        )
+
+        createMapboxNavigation(createRouteRefreshOptionsWithInvalidInterval(5_000))
+        mapboxNavigation.startTripSession()
+        val requestedRoutes = requestRoutes(enableRefresh = false)
+        mapboxNavigation.setNavigationRoutes(requestedRoutes)
+
+        delay(7000) // refresh interval + accuracy
```

Comment:
I don't think it can be tested properly in unit tests. They test many parts of the SDK along with the server interaction. While writing these tests I've discovered a couple of bugs and inconsistencies. What's more, they give you a feel of how the SDK works which might not have been so obvious before writing/reading them. I find them very useful and would like to keep them.
What exactly is the problem with additional 2 mins? Aren't the tests paralleled somehow?

Reply:
This is 2 minutes for a single test class. As we keep expanding at some point these types of granular addition will just unbearably extend the CI execution time. We're not at the critical point for the instrumentation tests job yet but we should be mindful of that as we go. If we can trim the execution down (for example, 7 seconds of a delay really shouldn't be necessary), let's do that.

> Aren't the tests paralleled somehow?

No, we're not doing [test sharding](https://firebase.google.com/docs/test-lab/android/instrumentation-test#sharding) yet but that idea has been floated in the past. Something to consider cc @SevaZhukov @dchernousov.

Reply:
I just mean that we shouldn't be held back from writing tests because they require some time to execute... In order not to be flaky at the same time. I don't know how to reduce the delay here:
1) I want to check states triggered by refresh on demand;
2) in order to do that I have to trigger refresh in-between planned attempts;
3) I have to wait for the first planned attempt;

If I make the interval 2-3 seconds instead of 5, code execution time will play a much bigger role in this test being stable.
In general, it's a common problem for tests that check that something is done with correct interval. And I think we should have such tests. For example, it would check that we don't do by mistake any unnecessary requests and so on.

Reply:
> I just mean that we shouldn't be held back from writing tests because they require some time to execute...

Of course not, test coverage always prevails over test execution time. I just want us to be mindful, that's all. If you think that we can't optimize execution time any further without impacting stability, then let's leave this as is.

Reply:
I redid some of the tests into Robolectric tests.